text/0000-multi-sig-term.md

```diff
@@ -0,0 +1,340 @@
+`RequireSignature kh`. Leafs of the form `RequireAllOf` or `ReuquireMOf 0` with
+an empty argument list of multi-sig terms are redundant in the sense that they
+evaluate to `True`. Leaf terms of the form `RequireAnyOf` or `RequireMOf m`
+with `m>0` with an empty list of multi-sig terms is also redundant and evaluates
+to `False`.
+
+### Examples
+
+Assume we have key hashes for different participants: `aliceKH`, `bobKH`,
+`carlKH` and `dariaKH`. Then the simplest multi-signature term requires
+only Alice's signatures and looks as follows:
+
+```haskell
+aliceOnly = RequireSignature aliceKH
+```
+
+A multi-signature term that requires Alice's or Bob's signature looks as
+follows:
+
+```haskell
+aliceOrBob = RequireAnyOf [RequireSignature aliceKH, RequireSignature bobKH]
+```
+
+A script requiring both signatures looks as follows:
+
+```haskell
+aliceAndBob = RequireAllOf [RequireSignature aliceKH, RequireSignature bobKH]
+```
+
+Finally for an 2 out of 3 scheme for Alice, Bob and Carl, the multi-signature
+term looks as follows:
+
+```haskell
+aliceBobCarl2oo3 = RequireMOf 2 [RequireSignature aliceKH, RequireSignature bobKH,
+RequireSignature carlHK]
+```
+
+The terms can be combined recursively, e.g., requiring a signature of Bob and
+Alice or of Carl and Daria is expressed as follows:
+
+```haskell
+aliceAndBobOrCarlAndDaria = RequireAnyOf
+    [ RequireAllOf [RequireSignature aliceKH, RequireSignature bobKH]
+    , RequireAllOf [RequireSignature carlKH, RequireSignature dariaKH]]
+```
+
+`RequireAnyOf [msigs]` and `RequireMOf 1 [msigs]` are logically equivalent, just
+as `RequireAllOf [msigs]` and `RequireMOf (length msisg) [msigs]`. From a point
+of readability and to a lesser part also for encoding size, it makes sense to
+have the special case terms for _any_ and _all_.
+
+# Technical explanation
+[technical-explanation]: #technical-explanation
+
+Multi-signature terms can be used in value addresses and stake addresses. They
+can be used to lock unspent transaction outputs, withdrawal accounts and
+delegation certificates.
+
+## Binary Representation
+
+The binary translation uses the standardized CBOR format. Each of the possible
+sub-terms is tagged with an identifier, standard list encoding with start and
+break tags is used for a list of sub-terms and key hashes are encoded as an
+array of data items with the length depending on the specific hash
+algorithm. Using `SHA256` this length is `2` bytes for length encoding and `32`
+bytes for the hash. In CBOR CDDL:
+
+```
+msig =
+       [0, keyhash]          ; The case: RequireSignature sig
+     / [1, [ *msig ]]        ; The case: RequireAllOf msigs
+     / [2, [ *msig ]]        ; The case: RequireAnyOf msigs
+     / [3, uint, [ *msig ]]  ; The case: RequireMOf m msigs
+
+keyhash = bytes .size 32
+```
+
+*Note* the encoding of `m` into a major type `0` single byte allows for a
+maximal value of `23` for `m` in a `m` out of `n` multi-sig scheme.
```

Comment:
I don't think we need to restrict it to the one-byte form. We just need to limit the overall script size.

Reply:
thanks, was left-over from an initial version that used raw bytes description only